File: backend/src/services/ai-ops/topology/stateMachine.ts
/**
 * 三态状态机
 *
 * pending → confirmed → stale → removed
 * 支持分层稳定性阈值（infrastructure vs endpoint）
 *
 * Property 5: 三态状态机转换正确性
 * Property 6: Stale 过期移除
 * Requirements: 3.5, 3.6, 3.7, 5.5
 */

import { NodeState, StabilityTier, TopologyDiscoveryConfig } from './types';

export interface StateMachineEntity {
  state: NodeState;
  confirmCount: number;
  missCount: number;
  lastSeenAt: number;
  stabilityTier?: StabilityTier;
}

/**
 * 获取确认阈值
 */
export function getConfirmThreshold(tier: StabilityTier, config: TopologyDiscoveryConfig): number {
  return tier === 'infrastructure' ? config.infraConfirmCount : config.endpointConfirmCount;
}

/**
 * 获取失效阈值
 */
export function getStaleThreshold(tier: StabilityTier, config: TopologyDiscoveryConfig): number {
  return tier === 'infrastructure' ? config.infraStaleThresholdCount : config.endpointStaleThresholdCount;
}

/**
 * 处理实体被发现（本轮采集中出现）
 * 返回更新后的状态
 */
export function onEntitySeen(entity: StateMachineEntity, tier: StabilityTier, config: TopologyDiscoveryConfig, now: number): StateMachineEntity {
  const confirmThreshold = getConfirmThreshold(tier, config);
  const newConfirmCount = entity.confirmCount + 1;

  if (entity.state === 'pending') {
    if (newConfirmCount >= confirmThreshold) {
      return { ...entity, state: 'confirmed', confirmCount: newConfirmCount, missCount: 0, lastSeenAt: now };
    }
    return { ...entity, confirmCount: newConfirmCount, missCount: 0, lastSeenAt: now };
  }

  if (entity.state === 'stale') {
    // stale → confirmed（重新被发现）
    if (newConfirmCount >= confirmThreshold) {
      return { ...entity, state: 'confirmed', confirmCount: newConfirmCount, missCount: 0, lastSeenAt: now };
    }
    return { ...entity, state: 'pending', confirmCount: newConfirmCount, missCount: 0, lastSeenAt: now };
  }

  // confirmed 状态：重置 missCount，更新 lastSeen
  return { ...entity, confirmCount: newConfirmCount, missCount: 0, lastSeenAt: now };
}

/**
 * 处理实体未被发现（本轮采集中缺失）
 * 返回更新后的状态，或 null 表示应从图中移除
 */
export function onEntityMissed(entity: StateMachineEntity, tier: StabilityTier, config: TopologyDiscoveryConfig, now: number): StateMachineEntity | null {
  const staleThreshold = getStaleThreshold(tier, config);
  const newMissCount = entity.missCount + 1;

  if (entity.state === 'confirmed') {
    if (newMissCount >= staleThreshold) {
      return { ...entity, state: 'stale', missCount: newMissCount };
    }
    return { ...entity, missCount: newMissCount };
  }

  if (entity.state === 'stale') {
    // 检查是否超过过期时间
    if (now - entity.lastSeenAt > config.staleExpiryMs) {
      return null; // 从图中移除
    }
    return { ...entity, missCount: newMissCount };
  }

  // pending 状态：超过过期时间则移除
  if (now - entity.lastSeenAt > config.staleExpiryMs) {
    return null;
  }
  return { ...entity, missCount: newMissCount };
}

File: opsevo-python/src/opsevo/services/topology/state_machine.py
"""
三态状态机 — pending → confirmed → stale → removed

支持分层稳定性阈值（infrastructure vs endpoint）

Requirements: 16.2
"""

from __future__ import annotations

from dataclasses import dataclass

from opsevo.services.topology.types import (
    NodeState, StabilityTier, TopologyDiscoveryConfig,
)


@dataclass
class StateMachineEntity:
    state: NodeState
    confirm_count: int
    miss_count: int
    last_seen_at: float
    stability_tier: StabilityTier = StabilityTier.INFRASTRUCTURE


def get_confirm_threshold(tier: StabilityTier, config: TopologyDiscoveryConfig) -> int:
    return config.infra_confirm_count if tier == StabilityTier.INFRASTRUCTURE else config.endpoint_confirm_count


def get_stale_threshold(tier: StabilityTier, config: TopologyDiscoveryConfig) -> int:
    return config.infra_stale_threshold_count if tier == StabilityTier.INFRASTRUCTURE else config.endpoint_stale_threshold_count


def on_entity_seen(
    entity: StateMachineEntity, tier: StabilityTier,
    config: TopologyDiscoveryConfig, now: float,
) -> StateMachineEntity:
    """处理实体被发现（本轮采集中出现）。"""
    threshold = get_confirm_threshold(tier, config)
    new_count = entity.confirm_count + 1

    if entity.state == NodeState.PENDING:
        new_state = NodeState.CONFIRMED if new_count >= threshold else NodeState.PENDING
        return StateMachineEntity(
            state=new_state, confirm_count=new_count, miss_count=0,
            last_seen_at=now, stability_tier=tier,
        )

    if entity.state == NodeState.STALE:
        if new_count >= threshold:
            new_state = NodeState.CONFIRMED
        else:
            new_state = NodeState.PENDING
        return StateMachineEntity(
            state=new_state, confirm_count=new_count, miss_count=0,
            last_seen_at=now, stability_tier=tier,
        )

    # confirmed: reset miss, update lastSeen
    return StateMachineEntity(
        state=NodeState.CONFIRMED, confirm_count=new_count, miss_count=0,
        last_seen_at=now, stability_tier=tier,
    )


def on_entity_missed(
    entity: StateMachineEntity, tier: StabilityTier,
    config: TopologyDiscoveryConfig, now: float,
) -> StateMachineEntity | None:
    """处理实体未被发现。返回 None 表示应从图中移除。"""
    stale_threshold = get_stale_threshold(tier, config)
    new_miss = entity.miss_count + 1

    if entity.state == NodeState.CONFIRMED:
        new_state = NodeState.STALE if new_miss >= stale_threshold else NodeState.CONFIRMED
        return StateMachineEntity(
            state=new_state, confirm_count=entity.confirm_count,
            miss_count=new_miss, last_seen_at=entity.last_seen_at, stability_tier=tier,
        )

    if entity.state == NodeState.STALE:
        if (now - entity.last_seen_at) * 1000 > config.stale_expiry_ms:
            return None
        return StateMachineEntity(
            state=NodeState.STALE, confirm_count=entity.confirm_count,
            miss_count=new_miss, last_seen_at=entity.last_seen_at, stability_tier=tier,
        )

    # pending
    if (now - entity.last_seen_at) * 1000 > config.stale_expiry_ms:
        return None
    return StateMachineEntity(
        state=NodeState.PENDING, confirm_count=entity.confirm_count,
        miss_count=new_miss, last_seen_at=entity.last_seen_at, stability_tier=tier,
    )
